lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -1062,6 +1062,92 @@ bool SessionState::deactivate_scheduled_static_rule(
   return true;
 }
 
+void SessionState::process_rule_installs(
+    std::vector<StaticRuleInstall> static_rule_installs,
+    std::vector<DynamicRuleInstall> dynamic_rule_installs,
+    RulesToProcess& to_activate, RulesToProcess& to_deactivate,
+    RulesToSchedule& to_schedule, SessionStateUpdateCriteria& session_uc) {
+  std::time_t current_time = std::time(nullptr);
+  std::string ip_addr      = config_.common_context.ue_ipv4();
+  std::string ipv6_addr    = config_.common_context.ue_ipv6();
+  for (const auto& rule_install : static_rule_installs) {
+    const std::string& rule_id = rule_install.rule_id();
+    RuleLifetime lifetime(rule_install);
+    if (is_static_rule_installed(rule_id)) {
+      // Session proxy may ask for duplicate rule installs.
+      // Ignore them here.
+      continue;
+    }
+    PolicyRule static_rule;
+    if (!static_rules_.get_rule(rule_id, &static_rule)) {
+      MLOG(MERROR) << "static rule " << rule_id
+                   << " is not found, skipping install...";
+      continue;
+    }
+
+    // If the rule should be deactivated already, deactivate just in case and
+    // continue
+    if (lifetime.exceeded_lifetime(current_time)) {
+      optional<uint32_t> op_version =
+          deactivate_static_rule(rule_id, session_uc);
+      if (op_version) {
+        to_deactivate.append_versioned_policy(static_rule, *op_version);
+      }
+      continue;
+    }
+    // If the rule should be active now, install
+    if (lifetime.is_within_lifetime(current_time)) {
+      uint32_t version = activate_static_rule(rule_id, lifetime, session_uc);
+      to_activate.append_versioned_policy(static_rule, version);
+    }
+    // If the rule is for future activation, schedule
+    if (lifetime.before_lifetime(current_time)) {
+      schedule_static_rule(rule_id, lifetime, session_uc);
+      to_schedule.push_back(
+          RuleToSchedule(STATIC, rule_id, ACTIVATE, lifetime.activation_time));
+    }
+    // Schedule deactivation time in the future
+    if (lifetime.should_schedule_deactivation(current_time)) {
+      to_schedule.push_back(RuleToSchedule(
+          STATIC, rule_id, DEACTIVATE, lifetime.deactivation_time));
+    }
+  }
+
+  for (auto& rule_install : dynamic_rule_installs) {
```

Comment:
why can't this be a different function from the above one, none of this loop applies to the previous loop?

Reply:
yeah I think it would be cleaner to just separate them out into `process_static_rule_installs` and `process_dynamic_rule_installs`